NitroxClient/Communication/TcpClient.cs

```diff
@@ -1,83 +1,96 @@
-        {
-            Log.Info("Initializing TcpClient...");
-            packetReceiver = packetManager;
-            IsConnected = false;
-        }
-
-        public void Start(string ipAddress)
-        {
-            try
-            {
-                IPAddress address = IPAddress.Parse(ipAddress);
-                IPEndPoint remoteEP = new IPEndPoint(address, PORT);
-
-                Socket socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
-                connection = new Connection(socket);
-                connection.Connect(remoteEP);
-
-                if (connection.Open)
-                {
-                    IsConnected = true;
-                    connection.BeginReceive(new AsyncCallback(DataReceived));
-                }
-            }
-            catch (Exception e)
-            {
-                Log.Debug("Unforeseen error when connecting: " + e.GetBaseException());
-            }
-        }
-
-        public void Stop()
-        {
-            connection.Close();
-            IsConnected = false;
-            Log.InGame("Disconnected from server.");
-        }
-
-        public void Send(Packet packet)
-        {
-            connection.SendPacket(packet, new AsyncCallback(packetSentSuccessful));
-        }
-
-        public void packetSentSuccessful(IAsyncResult ar) { }
-
-        private void DataReceived(IAsyncResult ar)
-        {
-            Connection connection = (Connection)ar.AsyncState;
-
-            foreach (Packet packet in connection.GetPacketsFromRecievedData(ar))
-            {
-                packetReceiver.PacketReceived(packet);
-            }
-
-            if (connection.Open)
-            {
-                connection.BeginReceive(new AsyncCallback(DataReceived));
-            }
-            else
-            {
-                Log.Debug("Error reading data from server");
-                Stop();
-            }
-        }
-    }
-}
+﻿using System;
+using System.Net;
+using System.Net.Sockets;
+using NitroxClient.Communication.Abstract;
+using NitroxModel.Logger;
+using NitroxModel.Packets;
+using NitroxModel.Tcp;
+
+namespace NitroxClient.Communication
+{
+    public class TcpClient : IClient
+    {
+        private readonly DeferringPacketReceiver packetReceiver;
+        private int port;
+        private Connection connection;
+
+        public bool IsConnected { get; private set; }
+
+        public TcpClient(DeferringPacketReceiver packetManager)
+        {
+            Log.Info("Initializing TcpClient...");
+            packetReceiver = packetManager;
+            IsConnected = false;
+        }
+
+        public void Start(string ipAddress)
+        {
+            try
+            {
+                //If Ip address includes a port use the included port
+                port = 11000;
+                string[] splitIp = ipAddress.SplitByChar(':');
+
+                for (int i = 0; i < splitIp.Length; i++)

```

Comment:
Such a complicated loop is not necessary here, and it'll set `ipAddress = splitIp[0];` on every iteration. Consider doing something like this instead:

```cs
Validate.IsTrue(!string.IsNullOrEmpty(ipAddress));
string[] splitIp = ipAddress.Split(':');
ipAddress = splitIp[0];
if (splitIp.Length > 1)
{
    port = int.Parse(splitIp[1]);
}
```

... and eventually throw an error if more than one colon is found.

We might want to check if .NET includes such an `address:port` parser already (returning an `IPEndPoint` combining the two).

Reply:
```C#
if (splitIp.Length > 1)
{
    port = int.Parse(splitIp[1]);
    if (splitIp.Length > 2)
    {
        throw new Exception("Multiple ports detected!");
    }
}
```

Something like this should help prevent errors with multiple colons